Added support for 5 Channel S0 USB Pulse meter (original: 5-kanaals S0 Pulse Meter op USB)

Lurchi70 opened the pull request:
Implementation a new device protocol: sos_s0 

The 5 Channel S0 USB Pulse meter is a device based on Arduino Micro. It collects and buffers S0 impulse from up to five S0 devices and sends the retrieved S0-IMpulses within a 10 second timeframe to the serial port (e.g. /dev/ttyACM0) available via standard USB driver. The protocol is documented by the vendor.
The device is available here: https://www.sossolutions.nl/5-kanaals-s0-pulse-meter-op-usb

This change adds a new protocol "sos_s0" to vzlogger. The implementation opens the device port, applies the hardcoded baud rate and parity setting (9600/7/n/1) and reads the data from the serial device. 
The read data is provided to volkzaehler middleware by the usual ways.

Lurchi70 committed:
Support SOS S0 Pulse Meter via USB

Adding support for S0 pulse meter from https://www.sossolutions.nl
"5-kanaals S0 Pulse Meter op USB"

Lurchi70 committed:
Support SOS S0 Pulse Meter via USB

Adding support for S0 pulse meter from https://www.sossolutions.nl
"5-kanaals S0 Pulse Meter op USB"

Lurchi70 committed:
Merge branch 'master' of https://github.com/Lurchi70/vzlogger

# Conflicts:
#	include/protocols/MeterSOS_S0.hpp
#	src/protocols/MeterSOS_S0.cpp

Lurchi70 committed:
Add support for 'send_zero' option to sos_s0 protocols

J-A-U commented:
Interesting hardware you got there.

You might add this also to ./etc/vzlogger_generic.schema.json , for generated configuration.

Lurchi70 committed:
Several smaller changes

++ aligned source code with clang-format
++ modified vzlogger_generic.schema.json for generated configuration
++ added send_zero parameter to sample conf file

Lurchi70 committed:
Changed misleading comment in sourcefile

Lurchi70 reviewed (commented):
Looks good now. Please Review.

J-A-U commented:
@r00t- bump version?

r00t- commented:
thanks for your contribution,
nice to see vzlogger used as a platform to support new meters,
will try to review the code next week.

Lurchi70 commented:
Hi,

 

any new about this? Did you find the time for a review?

 

Best regards

 

 

Von: Thorben T. ***@***.*** 
Gesendet: Sonntag, 2. Juli 2023 09:54
An: volkszaehler/vzlogger
Cc: Lurchi70; Author
Betreff: Re: [volkszaehler/vzlogger] Added support for 5 Channel S0 USB Pulse meter (original: 5-kanaals S0 Pulse Meter op USB) (PR #596)

 

thanks for your contribution,
nice to see vzlogger used as a platform to support new meters,
will try to review the code next week.

—
Reply to this email directly, view it on GitHub <https://github.com/volkszaehler/vzlogger/pull/596#issuecomment-1616430779> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/A6IX3FBDXCASK3OGZ7IQB7TXOESIPANCNFSM6AAAAAAZWZYSWI> .
You are receiving this because you authored the thread.Das Bild wurde vom Absender entfernt.Message ID: ***@***.***>

Lurchi70 committed:
Merge branch 'volkszaehler:master' into master

r00t- commented:
@Lurchi70:
is it a priority for you to have this merged?
imho, as long as there are few users, they can/will just find the MR and use the code,
and also it would be good to have more than one user trying a meter before it gets merged.
(this is the same for a meter i implemented myself: https://github.com/volkszaehler/vzlogger/pull/480 )

if you want to have it merged, i'd have some small requests:
* you based your meter on copied code from MeterD0, which is fine in itself, but i would suggest to first commit an unmodified copy, so it's easy to see (and extract) your actual new code. (as i did in https://github.com/volkszaehler/vzlogger/pull/480/commits => https://github.com/volkszaehler/vzlogger/pull/480/commits/2a33a923d90f78732472cb567297baba0b81a561 + https://github.com/volkszaehler/vzlogger/pull/480/commits/657b0ad13a8ba6c29be6548cf1c6ae3a68808fcf ), i could assisti if help is needed.
* consider those weird stray comments like "RW: added ack" or "snip from kernel end" that were copied long (would be fixed by the above suggestion)
* (in the long run, we should extract the serial-port handling code into some shared functions instead of duplicating the code in every serial meter, but that doesn't need to be done here.)
* (probably too much, and i didn't do it myself either - if few people have the hardware, that would make it even more important to have units tests for the code, so we can avoid it breaking unexpectedly with compiler updates or refactorings.)

r00t- commented on include/protocols/MeterSOS_S0.hpp:
```diff
@@ -0,0 +1,72 @@
+ * Plaintext protocol of S0 Pulse meter devices
+ *  --> https://www.sossolutions.nl/5-kanaals-s0-pulse-meter-op-usb
+ *
+ * The device is a USB s0 logger device.
```
maybe just a little bit more description of the communication, at least "sending  a text-based protocol over a usb-connected uart interface"?
maybe the usb-IDs used by the device or a udev rule to identify it.
